### PR TITLE
Nest braces support

### LINKED ISSUE #281: More intelligent nested brace/quotes completion
e.g., Type (" would get us (""), then keep typing something between quotes, and type another quotes and right brace should just move cursor to the right to the completed brace.

### PR DESCRIPTION
**Pull request by HuanhuanSunMSFT**
Should fix #281, #307 and #308

### CODE REVIEW

File: PowerShellTools.HostService/ServiceManagement/IntelliSense/PowershellIntelliSenseService.cs [hunk 1]
```diff
@@ -36,44 +37,10 @@ public long RequestTrigger
...
-
-                            // Reset trigger
-                            _requestTrigger = 0;
-                        }
-                        catch (Exception ex)
-                        {
-                            ServiceCommon.Log("Failed to retrieve the completion list per request due to exception: {0}", ex.Message);
-                        }
-                    });
+                    _requestTrigger = value;
```
Comment (ASayre) [inline PowerShellTools.HostService/ServiceManagement/IntelliSense/PowershellIntelliSenseService.cs]: Why is this being locked here?

File: PowerShellTools/Intellisense/AutoCompletionController.cs [hunk 2]
```diff
@@ -52,34 +52,59 @@ internal sealed class AutoCompletionController : IOleCommandTarget
             _editorOperations = editorOperations;
             _undoHistory = undoHistory;
             _serviceProvider = serviceProvider;
+            _autoCompleteCount = 0;
```
Comment (AndreSayreMSFT) [inline PowerShellTools/Intellisense/AutoCompletionController.cs]: Where is this being used?

File: PowerShellTools/Intellisense/AutoCompletionController.cs [hunk 3]
```diff
@@ -52,34 +52,59 @@ internal sealed class AutoCompletionController : IOleCommandTarget
...
+        }
+
         #region IOleCommandTarget implementation
 
         public int Exec(ref Guid pguidCmdGroup, uint nCmdID, uint nCmdexecopt, IntPtr pvaIn, IntPtr pvaOut)
         {
+            var command = (VSConstants.VSStd2KCmdID)nCmdID;
+
             if (VsShellUtilities.IsInAutomationFunction(_serviceProvider) ||
                 pguidCmdGroup != VSConstants.VSStd2K ||
```
Comment (ASayre) [inline PowerShellTools/Intellisense/AutoCompletionController.cs:87]: Can we pull some of these Gnarly ifs into their own private function for documentation/readability's sake?

**Comment by ASayre** [comment]
Are we able to have the PR's be realted bugs, instead of multiple issues all contained within this one PR and branch?

File: PowerShellTools.Test/IntelliSense/AutoCompletionControllerUnitTests.cs [hunk 4]
```diff
@@ -26,7 +26,7 @@ public void ShouldCompleteBrace()
             _mockedScript = @"MyFunc";
             _mockedCaret = 6;
 
-            int actual = autoCompletionController.ProcessKeystroke((uint)VSConstants.VSStd2KCmdID.TYPECHAR, '(');
+            int actual = autoCompletionController.ProcessKeystroke(VSConstants.VSStd2KCmdID.TYPECHAR, '(');
```
Comment (ASayre) [inline PowerShellTools.Test/IntelliSense/AutoCompletionControllerUnitTests.cs:29]: Do we have new unit tests to cover all these new scenarios and potential edge cases?

Comment (HuanhuanSunMSFT) [inline PowerShellTools.Test/IntelliSense/AutoCompletionControllerUnitTests.cs:29]: Just added.

**Comment by HuanhuanSunMSFT** [comment]
These are all nested braces related bugs. You saw those mess because I didn't merge changes from dev branch. Now you should see a clean PR :) @AndreSayreMSFT

**Comment by ASayre** [comment]
looks good :)
